### ISSUE TITLE
Thread safety issue

### ISSUE THREAD

**Issue by anmarcek** [expressmapper/ExpressMapper#40]
Dear Yuriy,

thank you for a great job with the mapper, I like it very much.
However I just stumpled upon a problem with thread safety (the problem might be elsewhere but at least a quick look at the mapper source codes gave me the impression it is about thread safety).

I am using the mapper version 1.1.2 and I get the exception "An item with the same key has already been added." randomly when I call

```
Mapper.Map<IEnumerable<ItemA>, List<ItemB>>(someItems)
```

I believe the issue lies in the fact that Mapper.Map is not enclosed in a lock statement (which is fine of course), but in this case the call to Map method at the end results in calling MappingService.PrecompileCollection method which definetely should be thread safe.

Here is the stack trace:
```
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at ExpressMapper.MappingServiceProvider.PrecompileCollection[T,TN]() in C:\SourceCode\ExpressMapper\ExpressMapper NET40\MappingServiceProvider.cs:line 86
   at ExpressMapper.MappingServiceProvider.Map[T,TN](T src, TN dest) in C:\SourceCode\ExpressMapper\ExpressMapper NET40\MappingServiceProvider.cs:line 192
   at ExpressMapper.Mapper.Map[T,TN](T src) in C:\SourceCode\ExpressMapper\ExpressMapper NET40\Mapper.cs:line 35
```

Feel free to let me know if you need more details.
Thank you in advance, Tony

**Comment by anisimovyuriy** [comment]
Dear Anton,

Thanks a lot for your valuable input!
It's really weird as in PrecompileCollection method has a lock statement in place. And it is the only point where "Shared State" while executing "Map" method is modified. I'll look into it for sure. For now I'd like to tell you about "PrecompileCollection" method. This method is located in the public interface you can use ```Mapper.PrecompileCollection<T,TN>()```
It has created for one purpose to precompile "direct" collection e.g. When you have a map like:
```Mapper.Register<Test, TestViewModel>()```
In the case if you try to map any collection "direct" type:
```Mapper.Map<IEnumerable<Test>, List<TestViewModel>>(src)```
Then PrecompileCollection<IEnumerable<Test>, List<TestViewModel>> executed implicitly and compiles iteration through the source collection and creating plus adding elements into the destination one.

What you can do right now and this is best practices inside of mapping registration precompile all your collection types in code which improves your performance at runtime. I'll try look into it and reproduce it and get back to you ASAP.

Thanks a lot!

**Comment by anmarcek** [comment]
Dear Yuriy,

you are right, precompiling the collection explicitly at app start does work as a workaround and solves the issue. Thank you :).

I have not downloaded the source codes, just checked them on Github which is not ideal but indeed the following hint may help you solve the issue:

Yes, Mapper.PrecompileCollection does use the lock properly. However, please, consider the following code execution:

```
Mapper.Map<T, TN>(T src) - no lock, but so far fine
MappingServiceProvider.PrecompileCollection<T, TN>() - no lock, but so far fine
DestinationMappingService.PrecompileCollection<T, TN>() - no lock and here comes the problem
```

The DestionationMappingService.PrecompileCollection method first checks if the calculated cacheKey exists within CollectionMappers (it is not there yet), then it precompiles the collection and at the end it tries to store it inside of the CollectionMappers which is a problem because in the meantime another thread already accessed the same method for the same collection - same cache key.

I might be totally wrong as reading source codes on Github is quite lame approach but it may help ;).

Best regards,
Anton

**Comment by anisimovyuriy** [comment]
Fixed - will be included in 1.5 release
